aiven/resource_project_vpc.go

```diff
@@ -154,13 +178,61 @@ func (w *ProjectVPCActiveWaiter) RefreshFunc() resource.StateRefreshFunc {
 
 // Conf sets up the configuration to refresh.
 func (w *ProjectVPCActiveWaiter) Conf() *resource.StateChangeConf {
-	state := &resource.StateChangeConf{
-		Pending: []string{"APPROVED"},
-		Target:  []string{"ACTIVE"},
-		Refresh: w.RefreshFunc(),
-	}
-	state.Delay = 10 * time.Second
-	state.Timeout = 4 * time.Minute
-	state.MinTimeout = 2 * time.Second
-	return state
+	return &resource.StateChangeConf{
+		Pending:    []string{"APPROVED", "DELETING", "DELETED"},
+		Target:     []string{"ACTIVE"},
+		Refresh:    w.RefreshFunc(),
+		Delay:      10 * time.Second,
+		Timeout:    4 * time.Minute,
+		MinTimeout: 2 * time.Second,
+	}
+}
+
+// ProjectVPCDeleteWaiter is used to wait for VPC been deleted.
+type ProjectVPCDeleteWaiter struct {
+	Client  *aiven.Client
+	Project string
+	VPCID   string
+}
+
+// RefreshFunc will call the Aiven client and refresh it's state.
+func (w *ProjectVPCDeleteWaiter) RefreshFunc() resource.StateRefreshFunc {
+	return func() (interface{}, string, error) {
+		vpc, err := w.Client.VPCs.Get(w.Project, w.VPCID)
+		if err != nil {
+			// might be already gone after deletion
+			if err.(aiven.Error).Status == 404 {
+				return &aiven.VPC{}, "DELETED", nil
+			}
+
+			return nil, "", err
+		}
+
+		if vpc.State != "DELETING" && vpc.State != "DELETED" {
+			err := w.Client.VPCs.Delete(w.Project, w.VPCID)
+			if err != nil {
+				// VPC cannot be deleted while there are services migrating from it
+				// and coses 409 error API response
```

Comment:
Typo in `coses` but also the typical reason for failing to delete VPC is that services in it were deleted very recently and the cloud resources related to the VMs themselves have not yet been deleted, which prevents removing the VPC. Maybe "VPC cannot be deleted while there are services migrating from it" => "VPC cannot be deleted while there are services migrating from it or service deletion is still in progress".